phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
I'm just not a big fan of recursion depth limits.

Like what if I arbitrarily had a recursion depth limit 5 for File/ASN1.php? Some things go deeper than that so what am I to do in that situation? At that point File/X509.php might not even be able to parse the output and even if it could stuff might be missing and people would wonder why.

Having to set an arbitrary recursion depth limit is just asking for problems imho. It's just a non-intuitive setting that no one is ever going to think of to set. And even if they did they wouldn't know what'd be an appropriate value to set it to. It'd just be a method you'd need to call with a random parameter and it'd just be one extra step that people could mess up on.

I suppose in this particular case it's not as important. It reduces the initial entropy but they'll still be entropy.

It looks like the issue on older verions of PHP is the `===`:

```php
<?php
$a = array();
$a[] = &$a;

echo $a === $a[0] ? 'match'  : 'no match';
```

That causes the error.

One thing that does seem to work...  `debug_zval_dump`. According to http://php.net/debug-zval-dump it's been available as early as PHP 4.2.

It works with recursive stuff:

```php
<?php
$a = array();
$a[] = &$a;

debug_zval_dump($a);
```

It also works with private __sleep()'s:

```php
<?php
class test {
  var $var = 'zz';

  private function __sleep() {
    return array('var');
  }
}

$a = new test();

debug_zval_dump($a);
```

The one disadvantage to it is that it does dump to stdout. Output buffering (eg. `ob_start()` / `ob_flush()`) would need to be used to capture the output to a string. And that poses problems for people who might already be utilizing output buffering.

Another possibility that occured to me...

```php
<?php
$a = array();
$a[] = &$a;

$a = (object) $a;
```
`var_dump` seems to suggest that that works but `$a->0` isn't a valid variable name. I can't even hit it with reflection (not that that'd work on PHP 4):


```php
<?php
$a = array();
$a[] = &$a;

$a = (object) $a;

$reflectionClass = new ReflectionObject($a);
echo $a === $reflectionClass->getProperty('0')->getValue() ? 'matches' : 'does not match';
```

So I guess the recursion depth is the best approach meh. And if we go that route the next question would, I guess, be this: what would be a good recursion depth? If we did 1 that'd make for much simpler serialization. eg.

```php
    function phpseclib_safe_serialize($arr)
    {
        if (is_object($arr)) {
            return '';
        }
        if (!is_array($arr)) {
            return serialize($arr);
        }
        foreach (array_keys($arr) as $key) {
            if (is_object($arr[$key]) || is_array($arr[$key])) {
                unset($arr[$key]);
            }
        }
        return serialize($safearr);
    }
```